### PR TITLE
perf(db): amortize per-insert retention COUNT(*) in event repositories (#2799)

### LINKED ISSUE #2799: perf(db): amortize per-insert retention COUNT(*) in event repositories
## Summary
Six event-telemetry repositories run a full-table `SELECT count(*)` on **every single insert** to decide whether to trim to their 10,000-row retention cap. The `cleanupInProgress` boolean only prevents *concurrent* cleanups; it does not throttle the count, so the scan fires once per recorded event. Under bursty telemetry (network capture, log/OS event streaming), this doubles the query count per insert for a decision that is only meaningful once every ~10k rows. Found via a DB concurrency/migration audit (2026-07-01), finding P5count, priority P2, verified against source.

## Background & history
- Introduced: commit `d20e1f0e` (2026-03-16), [PR #1535](https://github.com/kaeawc/auto-mobile/pull/1535) — "feat: network/log/OS/custom telemetry SDK with IDE dashboard". This landed the `recordX` + `cleanupIfNeeded` retention pattern for the telemetry repos.
- The `cleanupInProgress` re-entrancy guard was added the next day in `3a7cc7417` (2026-03-17) — it guards against overlapping cleanups but does nothing to reduce how often the count runs.
- `networkEventRepository.ts` picked up the same shape via [PR #1624](https://github.com/kaeawc/auto-mobile/pull/1624) (2026-03-28).
- Why it looks like this: retention was written as a simple "check-then-trim after each write." That is correct and cheap to reason about, but it puts a scan on the hot insert path. The repo already has a better precedent for amortized retention (see below).

## The problem
Each `recordX` fires `cleanupIfNeeded` unawaited, which unconditionally runs a `count(*)` over the whole table:

```ts
// src/db/networkEventRepository.ts:34
export async function recordNetworkEvent(input, db?): Promise<number> {
  const result = await getDb(db).insertInto("network_events").values({...})
    .returning("id").executeTakeFirstOrThrow();
  cleanupIfNeeded(db);   // fire-and-forget, unawaited
  return result.id;
}
```

```ts
// src/db/networkEventRepository.ts:173
async function cleanupIfNeeded(db?): Promise<void> {
  if (cleanupInProgress) { return; }          // prevents concurrent cleanup, NOT the count
  cleanupInProgress = true;
  try {
    const d = getDb(db);
    const count = await d.selectFrom("network_events")
      .select(d.fn.countAll().as("count"))     // full-table scan on EVERY insert
      .executeTakeFirstOrThrow();
    if (Number(count.count) > RETENTION_MAX_ROWS) {
      const cutoff = await d.selectFrom("network_events")
        .select("timestamp").orderBy("timestamp", "desc")
        .offset(RETENTION_MAX_ROWS).limit(1).executeTakeFirst();
      if (cutoff) {
        await d.deleteFrom("network_events").where("timestamp", "<", cutoff.timestamp).execute();
      }
    }
  } catch {
    // best-effort cleanup   ← swallowed empty catch, no log (violates CLAUDE.md error convention)
  } finally {
    cleanupInProgress = false;
  }
}
```

Two issues:
1. **Per-insert scan.** The `count(*)` runs on every insert even though the trim decision only flips once the table crosses 10k rows. All of this runs inside the single daemon on the one `bun:sqlite` connection guarded by the async mutex in `src/db/bunSqliteDialect.ts`, so each retention count is another mutex-serialized query competing with real work on the hot path.
2. **Swallowed empty catch.** The `catch {}` block logs nothing, which violates the error-handling convention in `CLAUDE.md` (log-and-continue must log at `debug` with a one-line reason). A retention failure is currently invisible.

The `cleanupInProgress` guard does not help here: because `cleanupIfNeeded` is fire-and-forget and awaits between the boolean flip and the count, it merely drops *overlapping* runs — steady-state each insert still schedules its own count.

Identical code exists in five sibling repos:
- `src/db/logEventRepository.ts:75`
- `src/db/osEventRepository.ts:72`
- `src/db/navigationEventRepository.ts:72`
- `src/db/layoutEventRepository.ts:84`
- `src/db/storageEventRepository.ts:101`

(`failureAnalyticsRepository.ts:851` and `testExecutionRepository.ts:305` use a related retention guard but with a different call shape — fix the six `cleanupIfNeeded` repos first; align the other two if trivial.)

## Impact
Doubles the query count on the telemetry insert path: every recorded network/log/OS/navigation/layout/storage event triggers an extra full-table `count(*)`, each serialized through the shared connection mutex. On a small table this is cheap; under sustained capture (e.g. network mocking sessions, log streaming) it is measurable insert-path overhead and mutex contention for a decision that is almost always "no, don't trim." Narrow severity — it is self-throttling and never produces wrong results — but it is a per-insert scan that is nearly free to remove. Secondary: retention failures are currently swallowed silently.

## Proposed fix

> **Direction note (per review — #2799 may be a symptom, not the root cause).** Devin argues the per-insert `count(*)` is downstream of a bigger gap: **there is no shared telemetry ingestion path.** The six event repos (`network/log/os/navigation/layout/storage`) are byte-for-byte copies of the same shape (one-row `INSERT` + per-insert `count(*)` + duplicated `cleanupInProgress`/`RETENTION_MAX_ROWS = 10_000`/`cleanupIfNeeded` + swallowed empty `catch {}`), with no shared helper in `src/db`, and ingest is one-row-at-a-time (`AndroidCtrlProxyClient.handleWebSocketMessage` awaits `record*Event` per WS message; each `INSERT` blocks the single connection, `stmt.run()` synchronous — `bunSqliteDialect.ts:180`). The recommended direction is a **shared batched ingestion path in `TelemetryRecorder`** (the singleton choke point all six funnel through, `TelemetryRecorder.ts:29-43`): buffer inbound telemetry, flush on a short interval (mirror `performanceAuditRepository.startPeriodicPruning`'s `defaultTimer.setInterval(...).unref()`), and emit a multi-row `INSERT ... VALUES (...),(...)` per flush — turning N single-row auto-commits into one commit. A single shared `EventRetentionPolicy` (interface + fake per CLAUDE.md) then replaces all six copy-pasted `cleanupIfNeeded`s and runs on the flush tick. This subsumes #2799, retires much of #2787's per-single-row-commit fsync motivation, and addresses part of #2801's per-row body bloat. **Caveat:** batching changes durability/latency semantics (a crash loses the unflushed buffer; a subscribe-time backfill could miss the last few ms) — acceptable for best-effort telemetry, but a real design decision that deserves its own "shared batched telemetry ingestion" workstream/issue rather than being smuggled under "amortize a COUNT." If batching lands, **Option A below (per-repo `insertsSinceCleanup` counter) becomes the wrong shape** — it adds a seventh copy of per-repo state to six files we'd rather delete. Option B (shared timer) is the right seed and should be pushed up into the shared policy rather than duplicated per repo.

Stop counting on every insert. Two low-risk options; prefer whichever fits the module's lifecycle:

> **Preferred idiom (per review — reuse the proven amortized probe, not a new counter).** Sam verified that three sibling repos — `failureAnalyticsRepository.ts:861`, `performanceAuditRepository.ts:319`, `testExecutionRepository.ts:325` — already avoid the hot-path `count(*)` by using an **amortized threshold probe**: `.offset(RETENTION_MAX_ROWS - 1).limit(1)` to find the Nth-newest row directly (an indexed walk, not a full scan). Prefer mirroring that idiom in the six telemetry repos over introducing a divergent per-module counter — it keeps one canonical retention pattern (CLAUDE.md "one canonical primitive per concern") and removes the `count(*)` full scan *without* introducing module-level mutable counter state (the same footgun as `cleanupInProgress`). The three config look-alikes are NOT missed by this issue — they already use the amortized pattern, so the "6 telemetry event repos" scope is verified **complete**.

**Option A — amortize with a per-table in-memory counter (minimal, no lifecycle wiring):**
```ts
// src/db/networkEventRepository.ts
const CLEANUP_CHECK_INTERVAL = 256; // run the count(*) at most 1/256 inserts
let insertsSinceCleanup = 0;

// in recordNetworkEvent, replace `cleanupIfNeeded(db);` with:
if (++insertsSinceCleanup >= CLEANUP_CHECK_INTERVAL) {
  insertsSinceCleanup = 0;
  cleanupIfNeeded(db);
}
```
Since the cap is 10k and the counter resets each check, a 256-insert interval bounds worst-case overshoot to well under the cap while eliminating ~99.6% of the scans.

**Option B — periodic unref'd timer (mirror existing precedent):** move retention off the insert path entirely, following `performanceAuditRepository.startPeriodicPruning` (`src/db/performanceAuditRepository.ts:372`), which uses `defaultTimer.setInterval(...)` + `.unref()` so it never blocks process exit and is fully fake-timer-testable.

Either way, **also fix the swallowed catch** per `CLAUDE.md`:
```ts
} catch (error) {
  logger.debug(`network_events retention cleanup failed: ${error}`); // best-effort; safe to skip this cycle
}
```

Constraints the fix must respect:
- Must NOT change query results — the same rows get pruned, just less often / off the hot path.
- Must keep retention eventually enforced (table cannot grow unbounded).
- Apply the same change consistently to all six repos.
- If Option B is chosen, expose a `stop`/`unref` so tests and shutdown stay clean (mirror `stopPeriodicPruning`).

## Acceptance / evaluation criteria
- [ ] `recordNetworkEvent` (and the five siblings) no longer issue a `count(*)` on every insert — a unit test asserts the retention count runs at most once per N inserts (Option A) or only on the timer tick (Option B).
- [ ] Retention still enforced: inserting more than `RETENTION_MAX_ROWS + N` rows results in the table being trimmed back to the cap (verifiable behavioral assertion against a fake/in-memory DB).
- [ ] The swallowed `catch {}` in all six `cleanupIfNeeded` bodies logs at `debug` with a one-line reason, per `CLAUDE.md`. Verified locations: `networkEventRepository.ts:199`, `logEventRepository.ts:101`, `osEventRepository.ts:98`, `navigationEventRepository.ts:98`, `layoutEventRepository.ts:110`, `storageEventRepository.ts:127`. **(per review)** Additionally sweep the *second* empty catch at `storageEventRepository.ts:47` (not the cleanup one) — no empty `catch {}` should remain in the six repos; each must log the swallow with a reason.
- [ ] **(per review)** `RETENTION_MAX_ROWS` is injectable so the retention test does NOT insert 10k real rows (that would blow the <100ms budget); lower the cap via injection and assert trim-to-N. Assert the retention count/probe query runs the expected number of times (spy on the injected `db` or a query-counting fake dialect), and assert the previously-swallowed error now logs (spy the logger).
- [ ] Unit tests follow repo conventions (interfaces + fakes + `FakeTimer` if Option B), run in <100ms, non-flaky; add/extend `test/db/networkEventRepository.test.ts` and cover at least one sibling repo.
- [ ] Perf note: measure before/after query count for a batch of M inserts (e.g. M=1000) and record the drop in `count(*)` executions in the PR description; confirm pruned-row output is unchanged.
- [ ] `bun test`, `turbo run lint build test` green.

## References
- `src/db/networkEventRepository.ts:64` (unawaited call), `src/db/networkEventRepository.ts:173-204` (`cleanupIfNeeded`)
- `src/db/logEventRepository.ts:75`, `src/db/osEventRepository.ts:72`, `src/db/navigationEventRepository.ts:72`, `src/db/layoutEventRepository.ts:84`, `src/db/storageEventRepository.ts:101`
- Precedent: `src/db/performanceAuditRepository.ts:372-387` (`startPeriodicPruning`)
- Concurrency model: `src/db/bunSqliteDialect.ts`
- Finding P5count, DB concurrency/migration audit 2026-07-01

## Review findings incorporated

**Correctness / testability (Sam Whitfield)**
- The "6 event repos" scope is **verified complete**: grep of `cleanupInProgress`/`countAll`/`RETENTION` across `src/db/` returns 9 repos, but the extra three (`failureAnalyticsRepository.ts:861`, `performanceAuditRepository.ts:319`, `testExecutionRepository.ts:325`) already use the amortized `.offset(RETENTION_MAX_ROWS - 1).limit(1)` probe and do NOT have the per-insert `count(*)` bug. The six telemetry event repos are the only ones affected.
- Prefer mirroring the existing `offset(N-1).limit(1)` idiom over a new per-module counter, keeping one canonical retention pattern (CLAUDE.md) and avoiding a module-level mutable counter (same starvation/drift footgun as `cleanupInProgress`). The probe is an indexed walk, not a full scan.
- Empty `catch {}` is a real CLAUDE.md violation in all six repos; there is a **second** empty catch at `storageEventRepository.ts:47` beyond the cleanup one — sweep it too. Added an acceptance criterion.
- Make `RETENTION_MAX_ROWS` injectable so tests trim at a low cap in <100ms; assert probe-query count and that the previously-swallowed error now logs.

**Direction (Devin Osei)**
- #2799 is likely a **symptom**, not root cause: the six event repos are near-identical copy-paste with no shared ingestion abstraction, and ingest is one-row-at-a-time (each `INSERT` its own auto-commit through the single mutex-guarded connection). Recommends a shared **batched ingestion path in `TelemetryRecorder`** (buffer + multi-row `INSERT` + a single shared `EventRetentionPolicy`), which also addresses much of #2787's fsync-per-commit motivation and part of #2801's per-row body bloat.
- If batching lands, **Option A (per-repo counter) is the wrong shape** (a seventh copy of per-repo state); Option B (shared timer) is the right seed, pushed up into the shared policy. Batching's durability/latency trade-off deserves its own workstream/issue.

**Cross-links:** #2787 (synchronous=NORMAL / fsync-per-commit), #2788 (composite indexes for retention/read scans — independent additive win), #2801 (truncate backfill bodies), plus a potential shared-batched-telemetry-ingestion workstream.

### PR DESCRIPTION
**Pull request by kaeawc**
## Summary

The six telemetry event repositories (`network / log / os / navigation / layout /
storage EventRepository.ts`) ran a retention `SELECT count(*)` inside
`cleanupIfNeeded` on **every insert** to gate a 10k-row trim. This PR **amortizes**
that scan (issue #2799, Option A): it now runs at most once per
`CLEANUP_CHECK_INTERVAL` (256) inserts via a per-table insert counter, so 255/256
of the fire-and-forget retention calls are pure synchronous no-ops that never touch
the DB or the connection mutex.

Closes #2799.

## Why not "just replace count(\*) with an indexed offset probe"

That was the first approach here (and the reviewer-suggested "cheap indexed walk"),
but an adversarial perf review **benchmarked in bun:sqlite** and disproved the
premise:

| table rows | `count(*)` | `offset(cap).limit(1)` probe | ratio |
|-----------:|-----------:|-----------------------------:|------:|
| 10,001     | **0.58 µs**| **33.2 µs**                  | ~57x  |

`EXPLAIN QUERY PLAN` shows both use the covering `idx_<table>_timestamp`, but
`count(*)` compiles to the **page-granular `Count` opcode** (≈ O(pages)), while the
probe steps **once per index entry** to reach `OFFSET cap` (O(cap)). So swapping in
the probe made the hot path ~57x *slower* per scan, not faster. The `count(*)` is
the cheaper query; the right fix is to run it **less often**, not replace it.

## The change

```ts
const CLEANUP_CHECK_INTERVAL = 256;
let insertsSinceCleanup = 0;

export async function cleanupIfNeeded(db?, maxRows = RETENTION_MAX_ROWS, checkInterval = CLEANUP_CHECK_INTERVAL) {
  if (++insertsSinceCleanup < checkInterval) { return; }   // no DB, no guard — pure no-op
  insertsSinceCleanup = 0;
  if (cleanupInProgress) { return; }
  cleanupInProgress = true;
  try {
    // cheap page-granular count(*) gate; O(cap) offset probe only when actually over cap
    ...unchanged count-gated trim...
  } catch (error) {
    logger.warn(`<table> retention cleanup failed: ${error}`, error);   // was bare catch {}
  } finally { cleanupInProgress = false; }
}
```

- The counter is bumped **synchronously** at the top (before any `await`), so even
  though `recordX` dispatches `cleanupIfNeeded` fire-and-forget, retention still
  fires deterministically every N inserts — and the skipped calls return before
  touching `cleanupInProgress`, so there is no guard race.
- Pruned-row output is **unchanged**: when the gate fires it runs the identical
  count-gated `offset(maxRows)` cutoff + `timestamp < cutoff` delete as before.
- Worst-case overshoot is bounded to `cap + checkInterval` (10,256) rows —
  negligible against the 10k cap, and retention is still eventually enforced.

## Measured effect (M = 1000 inserts, in-memory)

- `count(*)` executions: **1000 → 3** (fires at inserts 256/512/768).
- Per-insert retention cost on the other 997 inserts: a counter increment + compare
  (no query, no mutex-held segment). This is a genuine reduction in queries
  serialized through the single `bun:sqlite` connection — the issue's actual motivation.

## Also in this PR

- **Injectable cap + interval.** `cleanupIfNeeded(db?, maxRows, checkInterval)` is
  exported so tests verify amortization and trim behavior at a low interval/cap in
  <100ms (no 10k rows, no real timers). `recordX` still calls it fire-and-forget
  with the defaults; behavior is unchanged.
- **Logged swallow (CLAUDE.md).** The five repos that still had a bare `} catch {}`
  in `cleanupIfNeeded` now `logger.warn(...)` with a one-line reason, matching the
  already-landed `storageEventRepository`. This overlaps #2894 by construction (the
  rewrite touches the same body); **#2894 can be closed as resolved for these repos.**

## Testing

New `test/db/eventRetention.test.ts` (parametrized over all six repos, in-memory DB,
no real timers, <100ms each):

- **amortization** — `count(*)` runs at most once per `checkInterval`; the other
  N-1 calls issue **zero** SQL (SQL-spy via Kysely `log`).
- **output-preserving trim** — insert 12, cap 5 → retains newest `[7..12]`, prunes oldest.
- **under cap** — prunes nothing.
- **failure logged** — a cleanup against a destroyed DB is caught, logged
  (`logger.warn` spy, asserts table name), and never propagates.

`bun run lint`, `bun run build`, and `bun test test/db/` all green locally. (The
prior offset-probe commit also passed all CI jobs incl. Windows; this rework only
changes the retention gate + tests.)

## Deliberate scope decisions / deferred follow-ups (filed separately)

- Kept the point-fix per-repo rather than Devin's shared batched-ingestion
  `TelemetryRecorder` / `EventRetentionPolicy` — the issue explicitly defers batching
  (a durability/latency change) to its own workstream — filed as #3138.
- The six repos still keep the existing `offset(maxRows)` + timestamp-only delete
  (N+1 steady state), not the canonical `offset(N-1)` + id-tiebreak of the three
  config repos, because #2799 constrains the fix to "must NOT change query results."
  Convergence (which *does* change pruned output) is filed as #3137.
- A shared `cleanupIfNeeded` helper to collapse the six near-identical bodies is
  filed as #3136 (independent of batching).

### CODE REVIEW

**Comment by kaeawc** [comment]
### Adversarial review round (3 persona agents + perf benchmarking)

Ran three independent review agents (perf-skeptic DBA, correctness/regression, architecture/direction). Summary of findings and how they were addressed:

**🔴 Perf-skeptic (blocking, valid):** the original approach — replacing `count(*)` with an `offset(cap).limit(1)` probe — was *inverted*. Benchmarked in bun:sqlite: at 10,001 rows `count(*)` = 0.58 µs vs the probe = 33.2 µs (**~57× slower**), because `count(*)` is the page-granular `Count` opcode while the probe steps once per index entry to reach the offset. **Addressed:** reworked to the issue's Option A — amortize the (cheap) `count(*)` gate behind a 256-insert counter. Measured `count(*)` executions over 1000 inserts: **1000 → 3**, with 997 inserts doing zero retention I/O.

**🟡 Correctness/regression:** verified output-equivalence across all edge cases (under cap, at cap, cap+1, same-ms bursts) — identical. Only actionable item was the 15 ms real-timer test flush (flake vector). **Addressed:** the amortized-counter design makes the skipped calls pure synchronous no-ops (they return before any `await` or guard mutation), so the flush is gone entirely — tests are deterministic with no real timers.

**🟢 Architecture/direction:** point-fix over shared abstraction is the right call for this issue; deviations flagged honestly. **Addressed:** filed follow-ups #3136 (shared retention helper), #3137 (converge onto the canonical `offset(N-1)`+id-tiebreak idiom — changes pruned output so can't ride this PR), #3138 (Devin's shared batched-ingestion workstream). #2894 (logged-swallow catch) is resolved here by construction — commented there to close on merge.

File: test/db/eventRetention.test.ts [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+      }
+    },
+  });
+  await runMigrations(db as Kysely<unknown>);
+  return { db, sqls };
+}
+
+/** Lets detached fire-and-forget cleanups from `recordX` settle before assertions. */
+async function flushPendingCleanups(): Promise<void> {
+  await new Promise<void>(resolve => defaultTimer.setTimeout(() => resolve(), 15));
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply test/db/eventRetention.test.ts]: Good catch — resolved in the follow-up commit (e5bc3dbd6) before merge. The rework switched from the offset-probe to the amortized-counter design, where the counter check returns *before* any `await` or `cleanupInProgress` mutation, so `recordX`'s fire-and-forget call is a pure synchronous no-op on skipped inserts. That removed the race entirely: the 15 ms `flushPendingCleanups` sleep is gone, behavior tests seed rows then call `cleanupIfNeeded(db, cap, /*checkInterval*/ 1)` (fires deterministically regardless of the module counter), and no test uses a real timer.
